guide: DVC Experiments Overview (#2909)
iterative/dvc.org: content/docs/user-guide/experiment-management/experiments-overview.md

> # DVC Experiments Overview

Reviewer: it can be a good page (still not clear if need a separate one for this though, considering that we have index)

should we do a diagram here with the basic workflow?

should we include cleaning here?

Author: > a diagram for the workflow would be nice

Since finalizing that properly may involve including alternative paths (a sort of flow chart) and design work, I vote to make it a follow-up issue (tied to the Exp Versioning release cc @dberenbaum) so we can merge this, if the content is approved in general.

DVC Experiments are captures automatically by DVC when you [run them].

[run them]: /doc/user-guide/experiment-management/running-experiments

> ## Basic Workflow

Reviewer: Should we name the whole section as "Basic workflow" if that's what this section is about anyway

Author: > name the whole section as "Basic workflow" if that's what this section is about

There is another intention here actually, which comes before the workflow. From https://github.com/iterative/dvc.org/pull/2909#discussion_r740633633:

>2... document implementation details about experiments (Git custom refs, etc.).

I just didn't do it yet because I thought it's planned in https://github.com/iterative/dvc.org/issues/2768 and didn't want to further complicate this PR which again is starting to grow beyond the original scope in terms of file changes... Maybe I'll start a nested one just to show what it looks like ⌛

Author: > > document implementation details about experiments
>
> I'll start a nested one just to show what it looks like

@shcheklein @iesahin please see https://github.com/iterative/dvc.org/pull/3006/files for this purpose. If you approve both PRs I'll merge the nested one first and then this one.

# DVC Experiments Overview

DVC Experiments are captures automatically by DVC when you [run them]. Each
experiment creates and tracks a project variation based on the changes in your
<abbr>workspace</abbr>. Experiments preserve the latest commit in the current
branch (Git `HEAD`) as their parent or _baseline_.

<details>

### ⚙️ How does DVC track experiments?

Experiments are custom [Git references] (found in `.git/refs/exps`) with a
single commit based on `HEAD` (not checked out by DVC). Note that these commits
are not pushed to Git remotes by default (see `dvc exp push`).

</details>

> Experiments will have an auto-generated ID like `exp-bfe64` by default. A custom

Reviewer 2: `auto-generated ID` is actually `auto-generated name`. We discussed this in another PR, considering there are too many different things that can be called ID, I think it might be easier for the reader just to call these _name_.

Author: Agreed. Updated @iesahin

name can be given instead (using the `--name`/`-n` option of `dvc exp run`).

> ID or name can be used to reference experiments with `dvc exp` subcommands.

[run them]: /doc/user-guide/experiment-management/running-experiments
[git references]: https://git-scm.com/book/en/v2/Git-Internals-Git-References

## Basic workflow

`dvc exp` commands let you automatically track a variation to a committed
project version (baseline). You can create independent groups of experiments
this way, as well as review, compare, and restore them later. The basic workflow
goes like this:

- Modify <abbr>parameters</abbr> or other dependencies (input data, source code,

>   stage definitions, etc.) of committed stages.

Reviewer 2: Modifying parameters and running the experiments occur at the same time with  `--set-param` option.

Author: That's fine, this is a simplified and generalized description of the "basic workflow" for dvc exps.

- [Run experiments] with `dvc exp run` (instead of `repro`) to execute the
  pipeline. The results are reflected in your <abbr>workspace</abbr>, and
  tracked automatically.

> - Use [metrics](/doc/command-reference/metrics) to identify the best

Reviewer 2: `dvc exp show` is used to identify the best experiments. Metrics provide a limited functionality compared to `exp show`.

Author: `exp show` displays metrics.

Author: I changed the link to /doc/user-guide/experiment-management/comparing-experiments now. I think that should resolve your concern.

Author: Never mind, rolling that back since we specifically mention `exp show` in the next bullet...

Author: Going back to this @iesahin , I merged the 2 bullets so the concern should be resolved now, I believe.

  experiment(s).

> - Visualize and compare experiments with `dvc exp show` or `dvc exp diff`.

Reviewer 2: I think we shouldn't call _visualize_ for the work performed by `exp show` and `exp diff`. `dvc plots`, or the DVCLive HTML report can be used for visualization.

Author: Changed to "review".

  Repeat 🔄
- Make certain experiments [persistent] by committing their results to Git. This

>   cleans the slate so you can repeat the process later.

Reviewer 2: I think _cleaning the slate_ is a bit misleading, the experiments stay until we make Git branches from them, or delete them. Committing some experiments to Git doesn't delete the others.

Author: Updated.

# DVC Experiments Overview

DVC Experiments are captures automatically by DVC when you [run them]. Each
experiment creates and tracks a project variation based on the changes in your
<abbr>workspace</abbr>. Experiments preserve the latest commit in the current
branch (Git `HEAD`) as their parent or _baseline_, but do not form part of the
regular Git tree or workflow (unless you make them [persistent]).

<details>

### ⚙️ How does DVC track experiments?

Experiments are custom [Git references] (found in `.git/refs/exps`) with a

> single commit based on `HEAD` (not checked out by DVC). Note that these commits

Reviewer 3: Experiments with checkpoints have multiple commits.

> (not checked out by DVC)

What's not checked out by DVC? `HEAD`? The experiment?

Reviewer 3: It might be good to mention the motivation for using custom refs and some basic analogy or description of what they are. There are some good examples in [#2979#discussion_r757848241](https://github.com/iterative/dvc.org/pull/2979#discussion_r757848241).

Author: Clarified.

Author: Not sure we need motivation for implementation details.

Author: p.s. I expanded on the general motivation a little based on the link. See d2c01119.

are not pushed to Git remotes by default (see `dvc exp push`).

</details>

[run them]: /doc/user-guide/experiment-management/running-experiments

> [git references]: https://git-scm.com/book/en/v2/Git-Internals-Git-References

Reviewer 3: Should we link to Peter's blog post?

Author: Great idea! Done.

## Properties

DVC Experiments will have an auto-generated name like `exp-bfe64` by default. A
custom name can be given instead (using the `--name`/`-n` option of
`dvc exp run`). These names can be used to reference experiments in other
`dvc exp` subcommands.

All experiments created by DVC will be associated to the latest commit (Git
`HEAD`) at the time that they were run. This is called the experiment's
_baseline_.

## Basic workflow

> `dvc exp` commands let you automatically track a variation to a committed
> project version (baseline). You can create independent groups of experiments

Reviewer 3: ```suggestion
`dvc exp` commands let you automatically track a variation to a git commit.
You can create independent groups of experiments
```

Minor suggestion to make this a little more direct. I'm not sure it's clear what a "project version" is or what "baseline" means here.

Author: Yeah it's more direct but it focuses too much on Git. It's not about Git commits.

"baseline" is defined in the previous paragraph.

I just removed "committed" instead (for now). WDYT?

this way, as well as review, compare, and restore them later. The basic workflow
goes like this:

- Modify hyperparameters or other dependencies (input data, source code,

>   commands to execute, etc.). Leave these changes un-committed in Git.

Reviewer 3: It might help to discuss why we have this workflow and want to leave changes un-committed. In other experiment tracking tools, the workflow looks like:

1. Run the experiment.
2. Log the commit hash of the current HEAD, even though it doesn't include your dirty changes.
3. If everything is successful, save changes in a new commit.

This creates a confusing state where the experiment should really be associated with the second commit instead of the first.

It might be too much detail or inappropriate for the page, but maybe it can be summarized, or it might spark other ideas. This is probably a good point for the blog post but not sure if there's a place for it...

Author: > Similar to the discussion up in https://github.com/iterative/dvc.org/pull/2909#pullrequestreview-812329316

Yeah this needs more work. I'm hoping for now it's mergeable but I'll improve is as much as possible ⌛

Author: > good point for the blog post

Do you mean linking from here? It's probably more important to link from the blog post to our docs, I think.

Reviewer 3: > Do you mean linking from here?

No, just thinking out loud that this idea would be good to incorporate into the blog post.

- [Run experiments] with `dvc exp run` (instead of `repro`). The results are
  reflected in your <abbr>workspace</abbr>, and tracked automatically.
- Review and [compare] experiments with `dvc exp show` or `dvc exp diff`, using
  [metrics](/doc/command-reference/metrics) to identify the best one(s). Repeat
  🔄
- Make certain experiments [persistent] by committing their results to Git. This
  lets you repeat the process from that point.

[run experiments]: /doc/user-guide/experiment-management/running-experiments
[pipeline]: /doc/user-guide/project-structure/pipelines-files
[compare]: /doc/user-guide/experiment-management/comparing-experiments
[persistent]: /doc/user-guide/experiment-management/persisting-experiments

## Initialize DVC Experiments on any project

DVC Experiments features build on basic semantics of <abbr>DVC projects</abbr>.
This means that minimal formalities are required, such as codifying a pipeline

> with `dvc.yaml` (even if it has a single <abbr>stage</abbr> that represents your

Author: OK guys as agreed I've incorporated a section about `exp init` in this page as well. @dberenbaum if you could review [this one page](content/docs/user-guide/experiment-management/experiments-overview.md) that's be great.

> Unfortunately we ended with a large and mixed PR again, and now it's also tied to the Experiment Versioning release due to this change. So it'd be great to try and finalize it soon but if I need to split it once again I can. Lmk if any of you think so.

Reviewer 3: Wow, quick work getting this into the guide!

Author: Thanks! It's just a couple paragraphs 😬. It may need more detail but I'm not super familiar with the latest iteration of the feature ATM.

entire process). Another typical preparation step is to create or modify a

> structured <abbr>parameters</abbr> file.

Reviewer 3: Should we also mention metrics? I think parameters and metrics are most important, since they may require writing additional code to read/write files, and it might be good to address this directly to avoid confusion.

Author: Sure, added. PTAL

`dvc exp init` lets you onboard any existing data science project to use DVC
Experiments without having to worry bootstrapping DVC manually. It will prompt
you a few simple questions and create a simple `dvc.yaml` file as well as other
<abbr>metafiles</abbr> with sane default values. You can review these and commit

> them to Git to begin using DVC Experiments.

Reviewer 3: Consider moving this paragraph before the other one. Introduce the command to get started before explaining the requirements of DVC projects.

Author: Yeah I guess that's more direct. Done (required a few adjustments), PTAL 2

# DVC Experiments Overview

DVC Experiments are captured automatically by DVC when [run]. Each experiment
creates and tracks a variation of your data science project based on the changes
in your <abbr>workspace</abbr>.

Experiments preserve a connection to the latest commit in the current branch
(Git `HEAD`) as their parent or _baseline_, but do not form part of the regular
Git tree or workflow (unless you make them [persistent]). This prevents

> polluting Git namespaces and bloating the repo unnecessarily.

Reviewer 3: Another option might be to reference that it's explained in more detail below.

> polluting Git namespaces and bloating the repo unnecessarily.

Author: > this paragraph feels a little abstract

Updated a little (3af2f9ac)

> Do we want to avoid terms like ephemeral/lightweight branches

Yes, those terms would also require definition and they're an implementation detail anyway. We're trying to stick to what matters most from the user perspective.

> reference that it's explained in more detail below

The title of the hidden section which is right after this paragraph should already serve as a reference or call-out for more info, I think.

</details>

## Basic workflow

`dvc exp` commands let you automatically track a variation of a project version
(the baseline). You can create independent groups of experiments this way, as
well as review, compare, and restore them later. The basic workflow goes like
this:

- Modify hyperparameters or other dependencies (input data, source code,
  commands to execute, etc.). Leave these changes un-committed in Git.
- [Run experiments][run] with `dvc exp run` (instead of `repro`). The results
  are reflected in your <abbr>workspace</abbr>, and tracked automatically.
- Review and [compare] experiments with `dvc exp show` or `dvc exp diff`, using
  [metrics](/doc/command-reference/metrics) to identify the best one(s). Repeat
  🔄
- Make certain experiments [persistent] by committing their results to Git. This
  lets you repeat the process from that point.

[pipeline]: /doc/user-guide/project-structure/pipelines-files
[compare]: /doc/user-guide/experiment-management/comparing-experiments
[persistent]: /doc/user-guide/experiment-management/persisting-experiments

## Initialize DVC Experiments on any project

DVC Experiments build on basic semantics of <abbr>DVC projects</abbr>. This

> means that minimal formalities are required.

Reviewer 3: > This means that minimal formalities are required.

I'm not sure what this is intended to mean. I can read it to mean pretty much opposite things (either DVC doesn't have a lot of formalities or that DVC requires some formalities), so I find it confusing. Is it needed?

Author: Hmmmm how about "To use DVC Experiments you need a <abbr>DVC project</abbr> with a minimal
structure and configuration" ?

`dvc exp init` lets you quickly onboard an existing data science project to use
DVC Experiments, without having to worry about bootstrapping DVC manually. You
can either supply a `command` to execute your experiments or use the
`--interactive` flag (`-i`) to be prompted for that and other optional

> customizations.

Reviewer 3: I think pushing `-i` hard here is worthwhile since it is by far the easiest way I have found to use the command, especially for beginners. Do we even need to mention `supply a command to execute your experiments`?

> This creates a simple `dvc.yaml` file for you. It uses sane default locations

Reviewer 3: Should we explain what `dvc.yaml` is at all or at least mention that it creates a pipeline stage?

> This creates a simple `dvc.yaml` file for you. It uses sane default locations

Reviewer 3: Can we jump right into customizable locations instead of first focusing on sane defaults (I would hope/assume it has sane defaults!).

Author: Well, the latest version is pretty different, want to take a look? Currently I have to mention the defaults before how to customize them.

Also I added a hypothetical link to a change I'm making in https://github.com/iterative/dvc.org/pull/3071 that specifies the default project structure (borrowed from the blog post).

well as review, compare, and restore them later. The basic workflow goes like
this:

- Modify hyperparameters or other dependencies (input data, source code,
  commands to execute, etc.). Leave these changes un-committed in Git.
- [Run experiments][run] with `dvc exp run` (instead of `repro`). The results
  are reflected in your <abbr>workspace</abbr>, and tracked automatically.
- Review and [compare] experiments with `dvc exp show` or `dvc exp diff`, using
  [metrics](/doc/command-reference/metrics) to identify the best one(s). Repeat
  🔄
- Make certain experiments [persistent] by committing their results to Git. This
  lets you repeat the process from that point.

[pipeline]: /doc/user-guide/project-structure/pipelines-files
[compare]: /doc/user-guide/experiment-management/comparing-experiments
[persistent]: /doc/user-guide/experiment-management/persisting-experiments

## Initialize DVC Experiments on any project

DVC Experiments build on basic semantics of <abbr>DVC projects</abbr>. This
means that minimal formalities are required.

`dvc exp init` lets you onboard any existing data science project to use DVC

> Experiments without having to worry about bootstrapping DVC manually. This

Reviewer 2: The user still needs to run `dvc init` before `dvc exp init` though.

Author: Good point! I added that note to the `exp init` ref instead of complicating this text.

# DVC Experiments Overview

DVC Experiments are captured automatically by DVC when [run]. Each experiment
creates and tracks a variation of your data science project based on the changes
in your <abbr>workspace</abbr>.

Experiments preserve a connection to the latest commit in the current branch
(Git `HEAD`) as their parent or _baseline_, but do not form part of the regular
Git tree or workflow (unless you make them [persistent]). This prevents
polluting Git namespaces and bloating the repo unnecessarily.

<details>

### ⚙️ How does DVC track experiments?

Experiments are custom [Git references](/blog/experiment-refs) (found in
`.git/refs/exps`) with one or more commits based on `HEAD`. These commits are
hidden and not checked out by DVC. Note that these are not pushed to Git remotes
by default either (see `dvc exp push`).

</details>

> [run them]: /doc/user-guide/experiment-management/running-experiments

Reviewer 3: There is a workflow section with this link below. Is it also needed here?

Author: Fixed.

## Properties

DVC Experiments will have an auto-generated name like `exp-bfe64` by default. A
custom name can be given instead (using the `--name`/`-n` option of
`dvc exp run`). These names can be used to reference experiments in other
`dvc exp` subcommands.

All experiments created by DVC will be associated to the latest commit (Git
`HEAD`) at the time that they were run. This is called the experiment's

> _baseline_.

Reviewer 3: Is this section needed? Naming details doesn't seem that important, and the baseline is already explained above.

Author: Hmm good points. I moved the naming paragraph to the hidden section for now (having it here instead of in the Running Exps guide was one of the original goals of the overview page). Removed the duplicate "baseline" explanation. Thanks

this:

- Modify hyperparameters or other dependencies (input data, source code,
  commands to execute, etc.). Leave these changes un-committed in Git.
- [Run experiments] with `dvc exp run` (instead of `repro`). The results are
  reflected in your <abbr>workspace</abbr>, and tracked automatically.
- Review and [compare] experiments with `dvc exp show` or `dvc exp diff`, using
  [metrics](/doc/command-reference/metrics) to identify the best one(s). Repeat
  🔄
- Make certain experiments [persistent] by committing their results to Git. This
  lets you repeat the process from that point.

[run experiments]: /doc/user-guide/experiment-management/running-experiments
[pipeline]: /doc/user-guide/project-structure/pipelines-files
[compare]: /doc/user-guide/experiment-management/comparing-experiments
[persistent]: /doc/user-guide/experiment-management/persisting-experiments

## Initialize DVC Experiments on any project

DVC Experiments build on basic semantics of <abbr>DVC projects</abbr>. This
means that minimal formalities are required.

`dvc exp init` lets you onboard any existing data science project to use DVC

> Experiments without having to worry bootstrapping DVC manually. It will prompt

Reviewer 3: ```suggestion
Experiments without having to worry bootstrapping DVC manually. `dvc exp init -i` will prompt
```

Might want to somehow specify to use `--interactive/-i` if mentioning prompts.

Author: Forgot about `-i` ⌛

Reviewer 3: Actually, we probably want to focus primarily on `dvc exp init -i` throughout this section. That's really the beginner-friendly way to get started.

Author: @dberenbaum I think both a simple `command` and `-i` are equally beginner-friendly. Mentioned both now.

> Experiments without having to worry bootstrapping DVC manually. It will prompt

Reviewer 3: ```suggestion
Experiments without having to worry about bootstrapping DVC manually. It will prompt
```

> you wth a few simple questions and create a basic `dvc.yaml` file, as well as

Reviewer 3: ```suggestion
you with a few simple questions and create a basic `dvc.yaml` file, as well as
```

Author: Fixed.

> you wth a few simple questions and create a basic `dvc.yaml` file, as well as

Reviewer 3: The rest of the paragraph after "create a basic `dvc.yaml` file` might be unnecessary. `dvc.yaml` also probably is unnecessary to introduce at all. You could probably combine this sentenc with the first sentence of the following paragraph, like:

> `dvc exp init -i` will prompt you with a few simple questions and create a basic pipeline (even if it has a single stage that represents your entire process).

Author: Gone.

> other <abbr>metafiles</abbr> with sane default values. You can review these

Reviewer 3: I don't think it actually creates any other files. It also might not even create `dvc.yaml` if one already exists, but not sure that's an important distinction here.

Author: > I don't think it actually creates any other files

Hmmm so how does it work? I guess at this point I need to go and actually use it again 😅 

It just grabs defaults from a new config folder right? Checking ⌛

Reviewer 3: It uses either the paths provided in the prompts/arguments, the config file, or the defaults (in that priority) to generate a stage in `dvc.yaml`.

Author: Got it. Updated, PTAL

Author: > the config file

p.s. I didn't mention config files to avoid complicating here. In fact this isn't very well explained in the ref. either -- I'll look into that.

> files and commit them to Git to begin using DVC Experiments quickly.

Reviewer 3: You shouldn't even need to commit to Git.

Author: > shouldn't even need to commit

Yeah but... You usually want to? Or I guess it's up to the user... ⌛

Reviewer 3: True. Not a big deal either way.

Author: I put this in parenthesis now.
